src/ring_theory/dedekind_domain/integer_unit.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2022 David Kurniadi Angdinata. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Kurniadi Angdinata
+-/
+
+import ring_theory.dedekind_domain.adic_valuation
+
+/-!
+# `S`-integers and `S`-units of fraction fields of Dedekind domains
+
+This file defines the ring of `S`-integers and the group of `S`-units of the field of fractions of a
+Dedekind domain, to be specialised to the case of a number field or a function field separately.
+
+TODO:
+* notation for `S`-integers and `S`-units.
+* proof that `S`-integers is the intersection of valuation rings.
+* proof that `S`-units is the kernel of a map to a product.
+* proof that `∅`-integers is the usual ring of integers.
+* finite generation of `S`-units and Dirichlet's `S`-unit theorem.
+
+## Main definitions
+
+* `is_dedekind_domain.S_integer`: the ring of `S`-integers.
+* `is_dedekind_domain.S_unit`: the group of `S`-units.
+
+## Main statements
+
+* `is_dedekind_domain.S_integer_unit`: the units of the `S`-integers are the `S`-units.
+
+## References
+
+* [D Marcus, *Number Fields*][marcus1977number]
+* [J W S Cassels, A Frölich, *Algebraic Number Theory*][cassels1967algebraic]
+* [J Neukirch, *Algebraic Number Theory*][Neukirch1992]
+
+## Tags
+
+S-integer, S-unit
+-/
+
+open_locale non_zero_divisors
+
+namespace is_dedekind_domain
+
+noncomputable theory
+
+universes u v
+
+variables {R : Type u} [comm_ring R] [is_domain R] [is_dedekind_domain R] (K : Type v) [field K]
+  [algebra R K] [is_fraction_ring R K] (S : set $ height_one_spectrum R)
+
+/-- The subring of `S`-integers of `K`. -/
+@[simps] def S_integer : subring K :=
+{ carrier   := {x : K | ∀ v ∉ S, (v : height_one_spectrum R).valuation x ≤ 1},
+  mul_mem'  := λ _ _ hx hy v hv, by simp only [map_mul, mul_le_one₀ (hx v hv) (hy v hv)],
+  one_mem'  := λ _ _, by simp only [map_one, le_refl],
+  add_mem'  := λ _ _ hx hy v hv, v.valuation.map_add_le (hx v hv) (hy v hv),
+  zero_mem' := λ _ _, by simp only [map_zero, zero_le_one],
+  neg_mem'  := λ _ hx v hv, by simp only [valuation.map_neg, hx v hv] }
+
+/-- The subgroup of `S`-units of `Kˣ`. -/
+@[simps] def S_unit : subgroup Kˣ :=
+{ carrier  := {x : Kˣ | ∀ v ∉ S, (v : height_one_spectrum R).valuation (x : K) = 1},
+  mul_mem' := λ _ _ hx hy v hv, by rw [units.coe_mul, map_mul, hx v hv, hy v hv, one_mul],
+  one_mem' := λ _ _, map_one _,
+  inv_mem' := λ _ hx v hv, by rw [valuation.map_units_inv, hx v hv, inv_one] }
+
+/-- The subring of `S`-integers of `K` is an algebra over `R`. -/
+instance : algebra R (S_integer K S) :=
```

Comment:
This can probably be golfed significantly, but I'm not familiar with the `..` hack

Reply:
The `..` pattern is no hack! I recommend you learn it but I don't think it can help here. What you have looks good to me.